python/cudf/cudf/core/copy_types.py

```diff
@@ -13,128 +14,161 @@
-    if len(column) == 0:
-        # Any empty column is valid as a gather map
-        # This is necessary because as_column([]) defaults to float64
-        # TODO: we should fix this further up.
-        # Alternately we can have an Optional[Column] and handle None
-        # specially in _gather.
-        return GatherMap(
-            cast("NumericalColumn", column.astype(size_type_dtype)),
-            nrows,
-            nullify,
-        )
-    if column.dtype.kind not in {"i", "u"}:
-        raise IndexError("Gather map must have integer dtype")
-    if not nullify and check_bounds:
-        lo, hi = libcudf.reduce.minmax(column)
-        if lo.value < -nrows or hi.value >= nrows:
-            raise IndexError(f"Gather map is out of bounds for [0, {nrows})")
-    return GatherMap(cast("NumericalColumn", column), nrows, nullify)
-
-
-def as_boolean_mask(column: Any, nrows: int) -> BooleanMask:
-    """Turn a column into a boolean mask
-
-    This augments the column with information that it is valid as a
-    boolean mask for columns with a given number of rows
-
-    Parameters
-    ----------
-    column
-        The column to verify
-    nrows
-        the number of rows to verify against
-
-    Returns
-    -------
-    BooleanMask
-        New object wrapping the column bearing witness to its
-        suitability as a boolean mask for columns with matching row
-        count.
-
-    Raises
-    ------
-    IndexError
-        If the column is of unsuitable dtype.
-    """
-    column = cudf.core.column.as_column(column)
-    if column.dtype.kind != "b":
-        raise IndexError("Boolean mask must have bool dtype")
-    if (n := len(column)) != nrows:
-        raise IndexError(
-            f"Column with {n} rows not suitable "
-            f"as a boolean mask for {nrows} rows"
-        )
-    return BooleanMask(cast("NumericalColumn", column))
+    def __init__(self, column: Any, nrows: int):
+        """A representation of a column as a boolean mask.
+
+        This augments the column with information that it is valid as a
+        boolean mask for columns with a given number of rows
+
+        Parameters
+        ----------
+        column
+            The data to turn into a column to then verify
+        nrows
+            the number of rows to verify against
+
+        Returns
+        -------
+        BooleanMask
+            New object wrapping the column bearing witness to its
+            suitability as a boolean mask for columns with matching
+            row count.
+
+        Raises
+        ------
+        IndexError
+            If the column is of unsuitable dtype. Or does not have the
+            requested number of rows.
+        """
+        self.column = cudf.core.column.as_column(column)
+        self.nrows = nrows
+        if self.column.dtype.kind != "b":
+            raise IndexError("Boolean mask must have bool dtype")
+        if len(column) != nrows:
+            raise IndexError(
```

Comment:
I vote for these to be `TypeError` and `ValueError`, respectively. Similarly for other classes in this file.

The docs for `IndexError` say that if an index is not an integer, `TypeError` is raised.

My preference for a `ValueError` is that this occurs during the construction of a gather map with an invalid value, whereas typically I only see `IndexError` raised when an out-of-bounds access is being performed. Here, the access never actually occurs because of the validation. If the gather could attempt a disallowed access, then perhaps `IndexError` would be suitable for that case.

https://docs.python.org/3/library/exceptions.html#IndexError

Reply:
I think I picked `IndexError` because otherwise I need to catch `ValueError` and raise `IndexError` in (for example) `DataFrame.take` and `xxx.iloc[out-of-bounds-index]`. I can do that, but potentially it hides other problems

Reply:
But changed the first to TypeError.

Reply:
@wence- also mentioned offline that this aligns with pandas. Please align with pandas, I hadn't considered the impact there.